token/program-2022/src/instruction.rs

```diff
@@ -409,6 +418,102 @@ pub enum TokenInstruction {
+    ///
+    InitializeMintWithTransferFee {
+        /// Number of base 10 digits to the right of the decimal place.
+        decimals: u8,
+        /// The authority/multisignature to mint tokens.
+        mint_authority: Pubkey,
+        /// The freeze authority/multisignature of the mint.
+        freeze_authority: COption<Pubkey>,
+        /// Pubkey that must own any provided fee account
+        fee_account_owner: Pubkey,
+        /// Amount of transfer collected as fees, expressed as basis points of the
+        /// transfer amount
+        transfer_fee_basis_points: u16,
+        /// Maximum fee assessed on transfers
+        maximum_fee: u64,
+    },
+    /// Transfer, providing an explicit fee receiver
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner/delegate
+    ///   0. `[writable]` The source account. Must be an AccountWithTransferFee.
+    ///   1. `[]` The token mint. Must be a MintWithTransferFee.
+    ///   2. `[writable]` The destination account. Must be an AccountWithTransferFee.
+    ///   3. `[writable]` The fee receiver account. Must be owned by the mint's
+    ///      `fee_account_owner`.
+    ///   4. `[signer]` The source account's owner/delegate.
+    ///
+    ///   * Multisignature owner/delegate
+    ///   0. `[writable]` The source account.
+    ///   1. `[]` The token mint.
+    ///   2. `[writable]` The destination account.
+    ///   3. `[writable]` The fee receiver account. Must be owned by the mint's
+    ///      `fee_account_owner`.
+    ///   4. `[]` The source account's multisignature owner/delegate.
+    ///   5. ..5+M `[signer]` M signer accounts.
+    TransferWithFee {
+        /// The amount of tokens to transfer.
+        amount: u64,
+        /// Expected number of base 10 digits to the right of the decimal place.
+        decimals: u8,
+    },
+    /// Transfer withheld tokens to a fee account. Signed by either the source
+    /// account owner or the fee account owner.
+    ///
+    /// Succeeds for frozen accounts. <-- Does this make sense?
+    ///
+    /// Note: this could be made permissionless, but that may open up attack
+    /// opportunities.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   * Single owner/delegate
+    ///   0. `[writable]` The source account. Must be an AccountWithTransferFee.
+    ///   1. `[]` The token mint. Must be a MintWithTransferFee.
+    ///   2. `[writable]` The fee receiver account. Must be owned by the mint's
+    ///      `fee_account_owner`.
+    ///   3. `[signer]` The source account's owner or the mint's `fee_account_owner`
+    ///
+    ///   * Multisignature owner/delegate
+    ///   0. `[writable]` The source account.
+    ///   1. `[]` The token mint.
+    ///   2. `[writable]` The destination account.
+    ///   3. `[]` The source or destination account's multisignature owner/delegate.
+    ///   4. ..4+M `[signer]` M signer accounts.
+    TransferWithheld {
```

Comment:
Instead of `TransferWithFee`/`TransferWithheld`, what if we had a `HarvestFee` instruction that accepts N input `AccountWithTransferFee` or `MintWithTransferFee` accounts, plus a destination `AccountWithTransferFee` that receives the full `withheld_amount` from all input accounts.  Signed by the `fee_withdraw_authority`

Reply:
The idea of `TransferWithheld` was for the case of wanting to close an account with a withheld amount still stuck in it, but if we store all withheld amounts on the mint, then you're right: we'll only need a `HarvestFee` instruction that takes the mint and an account, signed by the withdraw authority.

Reply:
> And we'll also need a way to move withheld tokens out, whether that's through Burn or another instruction.

I was thinking `Burn(0)` before `Close` could be used to move the withheld amount to the mint before closing.  But practically perhaps that's no different than a new variant of `Close` that accepts a writable mint account

Reply:
Let's do a new `Close` instruction.  It would be more composable to have a separate instruction which just moves withheld funds to the mint, but it's probably more abusable / confusing.  We can call it `CloseAccount2`?

Reply:
`ClosePaymentAccount`?